pkcs11module.c

```diff
@@ -1297,6 +1297,231 @@ PHP_METHOD(Module, C_DigestFinal) {
     efree(digest);
 }
 
+ZEND_BEGIN_ARG_INFO_EX(arginfo_C_SignInit, 0, 0, 3)
+    ZEND_ARG_TYPE_INFO(0, session, IS_OBJECT, 0)
+    ZEND_ARG_TYPE_INFO(0, mechanism, IS_ARRAY, 0) /* ['mechanism' => N, 'param' => str */
+    ZEND_ARG_TYPE_INFO(0, key, IS_LONG, 0)
+ZEND_END_ARG_INFO()
+
+PHP_METHOD(Module, C_SignInit) {
+    CK_RV rv;
+    CK_MECHANISM Mechanism = {.mechanism = 0, .pParameter = NULL, .ulParameterLen = 0};
+    CK_OBJECT_HANDLE hKey;
+
+    zval *mechanism;
+    zend_long key;
+    zval *session;
+
+    ZEND_PARSE_PARAMETERS_START(3, 3)
+        Z_PARAM_OBJECT_OF_CLASS(session, ce_Pkcs11_Session)
+        Z_PARAM_ARRAY(mechanism)
+        Z_PARAM_LONG(key)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (zend_hash_num_elements(Z_ARRVAL_P(mechanism)) > 2) {
+        zend_throw_exception(zend_ce_exception, "Invalid mechanism argument", 0);
+	return ;
+    }
+
+    /*
+     * Mechanism.mechanism = mechanism['mechanism'];
+     * Mechanism.pParameter = Z_STRVAL_P(mechanism['param']);
+     * Mechanism.ulParameterLen = Z_STRLEN_P(mechanism['param']);
+     */
+    zend_string *k;
+    zval *value;
+    zend_long h;
+
+    ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(mechanism), h, k, value) {
+        if (k && (strcmp(ZSTR_VAL(k), "mechanism") == 0)) {
+            if (Z_TYPE_P(value) != IS_LONG) {
+                zend_throw_exception(zend_ce_exception, "Invalid mechanism['mechanism'] argument, expecting an integer", 0);
+                return ;
+            }
+            Mechanism.mechanism = (CK_MECHANISM_TYPE)Z_LVAL_P(value);
+            ZVAL_DEREF(value);
+        }
+        if (k && (strcmp(ZSTR_VAL(k), "param") == 0)) {
+            if (Z_TYPE_P(value) != IS_STRING) {
+                zend_throw_exception(zend_ce_exception, "Invalid mechanism['param'] argument, expecting a string", 0);
+                return ;
+            }
+            Mechanism.pParameter = Z_STRVAL_P(value);
+            Mechanism.ulParameterLen = (CK_ULONG)Z_STRLEN_P(value);
+            ZVAL_DEREF(value);
+        }
+    } ZEND_HASH_FOREACH_END();
+
+    if (Mechanism.mechanism == 0) {
+        zend_throw_exception(zend_ce_exception, "Invalid mechanism", 0);
+        return ;
+    }
```

Comment:
There is a mechanism object that should be used rather than accepting an associative array. Check pkcs11key.c at PHP_METHOD(Key, initializeSignature)

Reply:
Fixed by the commit: https://github.com/gamringer/php-pkcs11/pull/18/commits/331c6d4b8757d07dff59db68c8a400062c187171